File: packages/shared-types/src/lesson.ts
export type LessonType = "reading" | "quiz" | "coding" | string;
export type LessonDifficulty = "easy" | "medium" | "hard" | string;
export type LessonStatus = "not_started" | "in_progress" | "completed" | string;

export type LessonResponse = {
  id: string;
  course_id: string;
  title: string;
  description: string | null;
  lesson_type: LessonType;
  difficulty: LessonDifficulty;
  sort_order: number;
  tags: string[];
  max_score: number;
  pass_score: number;
};

export type LessonDetailResponse = LessonResponse & {
  content: Record<string, unknown>;
};

export type LessonProgressResponse = {
  lesson_id: string;
  status: LessonStatus;
  best_score: number;
  attempt_count: number;
};

export type LessonListResponse = {
  lessons: LessonResponse[];
  progress: Record<string, LessonProgressResponse>;
};

export type SubmitLessonRequest = {
  submitted_code?: string | null;
  answer?: Record<string, unknown> | null;
  time_spent_seconds?: number | null;
};

export type SubmitLessonResponse = {
  submission_id: string;
  is_correct: boolean;
  score: number;
  ai_review_task_id: string | null;
};

File: apps/backend/app/schemas/lesson.py
from pydantic import BaseModel


class LessonResponse(BaseModel):
    id: str
    course_id: str
    title: str
    description: str | None
    lesson_type: str
    difficulty: str
    sort_order: int
    tags: list[str]
    max_score: int
    pass_score: int

    model_config = {"from_attributes": True}


class LessonDetailResponse(LessonResponse):
    content: dict


class LessonProgressResponse(BaseModel):
    lesson_id: str
    status: str
    best_score: int
    attempt_count: int


class LessonListResponse(BaseModel):
    lessons: list[LessonResponse]
    progress: dict[str, LessonProgressResponse] = {}


class SubmitLessonRequest(BaseModel):
    submitted_code: str | None = None
    answer: dict | None = None
    time_spent_seconds: int | None = None


class SubmitLessonResponse(BaseModel):
    submission_id: str
    is_correct: bool
    score: int
    ai_review_task_id: str | None = None
